File: python/tblink_rpc_core/transport_json_socket.py
'''
Created on Aug 22, 2021

@author: mballance
'''
import asyncio
import sys

from tblink_rpc_core.json.param2json import Param2Json
from tblink_rpc_core.param_val_int import ParamValInt
from tblink_rpc_core.param_val_map import ParamValMap
from tblink_rpc_core.param_val_str import ParamValStr
from tblink_rpc_core.param_val_bool import ParamValBool
from tblink_rpc_core.param_val_vec import ParamValVec
from tblink_rpc_core.transport import Transport
import selectors
import json
from tblink_rpc_core.json.json2param import Json2Param
import select


class TransportJsonSocket(Transport):
    DEBUG_EN = False
    
    def __init__(self, conn):
        self.req_f = None
        self.rsp_f = None
        self.conn = conn
        self.sel = selectors.DefaultSelector()
        self.sel.register(
            self.conn, 
            selectors.EVENT_READ, 
            self._recv_data)
        
        loop = asyncio.get_event_loop()
        loop.add_reader(
            self.conn.fileno(), 
            self._recv_data)
        
        self.drain_pending = False
        self.recv_state = 0
        self.recv_size  = 0
        self.msgbuf_tmp = ""
        pass
    
    def init(self, req_f, rsp_f):
        self.req_f = req_f
        self.rsp_f = rsp_f
        
        # Now that we've been initialized, we can
        # begin monitoring the file descriptor
        pass
   
    def send_req(self, method, id, params) -> int:
        if TransportJsonSocket.DEBUG_EN:
            print("--> Python: send_req method=%s" % method, flush=True)
        
        msg = ParamValMap()
        msg["method"] = ParamValStr(method)
        msg["params"] = params
        msg["id"] = ParamValInt(id)
        
        data = Param2Json().json(msg)
        
        if TransportJsonSocket.DEBUG_EN:
            print("data: %s" % data)
        
        header = ("Content-Length: %d\r\n\r\n" % len(data)).encode()
        self.conn.send(header)
        self.conn.send(data.encode())
#        if not self.drain_pending:
#            asyncio.ensure_future(self._drain())
#            self.drain_pending = True

        if TransportJsonSocket.DEBUG_EN:
            print("<-- Python: send_req method=%s" % method, flush=True)
        
        return id 
    
    def send_rsp(self, id, result, error):
        if TransportJsonSocket.DEBUG_EN:
            print("--> Python: send_rsp id=%d" % id, flush=True)
        
        msg = ParamValMap()
        msg["id"] = ParamValInt(id)
        if result is not None:
            msg["result"] = result
        elif error is not None:
            msg["error"] = error
        else:
            raise Exception("Neither result nor error provided")
        
        data = Param2Json().json(msg)
        
        if TransportJsonSocket.DEBUG_EN:
            print("data: %s" % data)
        
        header = ("Content-Length: %d\r\n\r\n" % len(data)).encode()
        self.conn.send(header)
        self.conn.send(data.encode())

        if TransportJsonSocket.DEBUG_EN:
            print("<-- Python: send_rsp id=%d" % id, flush=True)
            
    def _recv_data(self):
        if TransportJsonSocket.DEBUG_EN:
            print("--> _recv_data", flush=True)

        data = self.conn.recv(1024)
        
        if len(data) == 0:
            raise Exception("Zero-size data received")
        
        if TransportJsonSocket.DEBUG_EN:
            print("  received %d bytes" % len(data))
        
        self.msgbuf_tmp += data.decode()
        
        while len(self.msgbuf_tmp) > 0:
            if TransportJsonSocket.DEBUG_EN:
                print("  msgbuf: %s" % self.msgbuf_tmp)
            if self.recv_state == 0:
                #****************************************************
                #* Receive header
                #****************************************************
                if self.msgbuf_tmp.startswith("Content-Length: "):
                    self.msgbuf_tmp = self.msgbuf_tmp[len("Content-Length: "):]
                    self.recv_state = 1
                else:
                    if len(self.msgbuf_tmp) < len("Content-Length: "):
                        # Wait for more data to arrive
                        break
                    else:
                        print("Error: junk data received: %s" % self.msgbuf_tmp)
                        self.msgbuf_tmp = ""
            elif self.recv_state == 1:
                #****************************************************
                #* Receive message length
                #****************************************************
                newline_idx = self.msgbuf_tmp.find('\n')
                if newline_idx != -1:
                    size_s = self.msgbuf_tmp[0:newline_idx]
                    self.msgbuf_tmp = self.msgbuf_tmp[newline_idx:]
                    self.recv_size = int(size_s.strip())
                    
                    ws=0
                    while ws < len(self.msgbuf_tmp) and self.msgbuf_tmp[ws] in (' ','\r','\t','\n'):
                        ws += 1
                        
                    if ws > 0:
                        self.msgbuf_tmp = self.msgbuf_tmp[ws:]
                    self.recv_state = 2
            elif self.recv_state == 2:
                #****************************************************
                #* Receive message body
                #****************************************************
                if len(self.msgbuf_tmp) >= self.recv_size:
                    body = self.msgbuf_tmp[0:self.recv_size]
                    self.msgbuf_tmp = self.msgbuf_tmp[self.recv_size:]
                    
                    msg = json.loads(body)
                    
                    if "method" in msg.keys():
                        # This is a request
                        id = int(msg['id'])
                        method = msg["method"]
                        params = Json2Param().param(msg["params"])
                        
                        self.req_f(method, id, params)
                    else:
                        # This is a response
                        id = int(msg["id"])
                        result = None
                        error = None
                        
                        if "result" in msg.keys():
                            result = Json2Param().param(msg["result"])
                        if "error" in msg.keys():
                            error = Json2Param().param(msg["error"])
                            
                        self.rsp_f(id, result, error)
                    
                    self.recv_state = 0
                else:
                    print("ran out of space")
                    break
                    
        
        if TransportJsonSocket.DEBUG_EN:
            print("<-- _recv_data", flush=True)

    def process_one_message(self):
#        if TransportJsonSocket.DEBUG_EN:
#            print("--> process_one_message", flush=True)
        try:
            ready_to_read, ready_to_write, in_error = select.select([self.conn], [], [self.conn])
                
#            print("process_one_message: %d %d %d" % (
#                len(ready_to_read), len(ready_to_write), len(in_error)))

            if len(in_error) > 0:
                raise Exception("disconnect")
            elif len(ready_to_read) > 0:
                self._recv_data()
                
            if len(in_error) > 0:
                print("in_error: ", flush=True)
        except Exception:
            print("disconnect")
            raise Exception("disconnect")
         
            
#        if TransportJsonSocket.DEBUG_EN:
#            print("<-- process_one_message", flush=True)
            
    async def process_one_message_a(self):
        if TransportJsonSocket.DEBUG_EN:
            print("--> process_one_message_a", flush=True)
        loop = asyncio.get_event_loop()
        
        events = self.sel.select(0)

        for key, mask in events:
            key.data()

        await asyncio.sleep(0)

        if TransportJsonSocket.DEBUG_EN:
            print("<-- process_one_message_a", flush=True)
            
#    async def _drain(self):
#        await self.writer.drain()
#        self.drain_pending = False
    
    def mkValBool(self, v):
        return ParamValBool(v)
    
    def mkValIntS(self, v):
        return ParamValInt(v)
    
    def mkValIntU(self, v):
        return ParamValInt(v)
    
    def mkValMap(self):
        return ParamValMap()

    def mkValStr(self, v):
        return ParamValStr(v)

    def mkValVec(self):
        return ParamValVec()
    

File: cpp/TransportJsonSocket.cpp
/*
 * SocketMessageTransport.cpp
 *
 *  Created on: Sep 21, 2020
 *      Author: ballance
 */

#include "TransportJsonSocket.h"

#include <string>
#include <stdio.h>
#include <string.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <signal.h>
#include <unistd.h>
#ifndef _WIN32
#include <sys/select.h>
#include <sys/wait.h>
#endif
#include "nlohmann/json.hpp"

#include "Debug.h"
#include "JsonParamValFactory.h"
#include "ParamValBool.h"
#include "ParamValInt.h"
#include "ParamValMap.h"
#include "ParamValStr.h"
#include "ParamValVec.h"

#undef EN_DEBUG_TRANSPORT_JSON_SOCKET

#ifdef EN_DEBUG_TRANSPORT_JSON_SOCKET
DEBUG_SCOPE(TransportJsonSocket)
#define DEBUG_ENTER(fmt, ...) \
	DEBUG_ENTER_BASE(TransportJsonSocket, fmt, ##__VA_ARGS__)
#define DEBUG_LEAVE(fmt, ...) \
	DEBUG_LEAVE_BASE(TransportJsonSocket, fmt, ##__VA_ARGS__)
#define DEBUG(fmt, ...) \
	DEBUG_BASE(TransportJsonSocket, fmt, ##__VA_ARGS__)
#else
#define DEBUG(fmt, ...)
#define DEBUG_ENTER(fmt, ...)
#define DEBUG_LEAVE(fmt, ...)
#endif

namespace tblink_rpc_core {

TransportJsonSocket::TransportJsonSocket(
		pid_t		pid,
		int32_t 	socket) :
	m_msgbuf(0), m_msg_state(0), m_msg_length(0),
	m_pid(pid), m_socket(socket) {
	m_id = 0;

	m_outstanding = 0;

	// TODO Auto-generated constructor stub
	m_msgbuf_idx = 0;
	m_msgbuf_max = 1024;
	m_msgbuf = new char[m_msgbuf_max];
}

TransportJsonSocket::~TransportJsonSocket() {
	if (m_msgbuf) {
		delete [] m_msgbuf;
	}
}

void TransportJsonSocket::init(
		const recv_req_f			&req_f,
		const recv_rsp_f			&rsp_f) {
	m_req_f = req_f;
	m_rsp_f = rsp_f;
}

void TransportJsonSocket::shutdown() {
	DEBUG_ENTER("shutdown");
	if (m_pid > 0) {
		int status = -1;
		for (uint32_t i=0; i<5; i++) {
			int ret = ::waitpid(m_pid, &status, WNOHANG);

			if (ret == 0) {
				DEBUG("Sleeping...");
				::sleep(1);
			} else if (ret == m_pid) {
				DEBUG("ended");
				m_pid = 0;
				break;
			}
		}

		if (m_pid > 0) {
			fprintf(stdout, "TODO: terminate process\n");

			for (uint32_t i=0; i<5; i++) {
				int ret = ::kill(m_pid, (i<3)?SIGTERM:SIGKILL);

				ret = ::waitpid(m_pid, &status, WNOHANG);

//				DEBUG("waitpid ret=") << ret;

				if (ret == m_pid) {
					break;
				} else {
					DEBUG("Sleeping...");
					::sleep(1);
				}
			}
		}
	}
}

int32_t TransportJsonSocket::process_one_message() {
	char tmp[1024];
	int32_t sz;
	int32_t ret = 0;

	fd_set rfds, efds;
	struct timeval tv;
	int retval;

	FD_ZERO(&rfds);
	FD_ZERO(&efds);
	FD_SET(m_socket, &rfds);
	FD_SET(m_socket, &efds);

	/* Wait up to five seconds. */

	ret = select(m_socket+1, &rfds, 0, &efds, 0);

	if (ret > 0) {
		if (FD_ISSET(m_socket, &efds)) {
			fprintf(stdout, "Exception\n");
			ret = -1;
		} else if (FD_ISSET(m_socket, &rfds)) {
			sz = ::recv(m_socket, tmp, sizeof(tmp), 0);
			if (sz > 0) {
				process_data(tmp, sz);
			} else {
				ret = -1;
			}
		}
	}

	return ret;
}

#ifdef UNDEFINED
int32_t TransportJsonSocket::poll(int timeout_ms) {
	char tmp[1024];
	int32_t sz;
	int32_t ret = 0;

	// Poll for data

	do {

		DEBUG("poll -- timeout=") << timeout_ms;

		if (timeout_ms > 0) {
			// Select to wait on data
			fd_set rfds;
			struct timeval tv;
			int retval;

			FD_ZERO(&rfds);
			FD_SET(m_socket, &rfds);

			/* Wait up to five seconds. */

			tv.tv_sec = timeout_ms/1000;
			tv.tv_usec = (timeout_ms%1000)*1000;

			retval = select(m_socket+1, &rfds, 0, 0, &tv);

			if (retval == -1) {
				fprintf(stdout, "Error: select %d\n", errno);
				ret = -1;
				break;
			} else if (retval == 0) {
				// No data
				DEBUG("Note: no data");
				timeout_ms = 0;
				break;
			} else {
				timeout_ms = 0;
			}
		}

		DEBUG_ENTER("poll recv");
		sz = ::recv(m_socket, tmp, 1024, 0);
		DEBUG_LEAVE("poll recv");

		DEBUG("sz=") << sz;

		if (sz <= 0) {
			DEBUG("sz=") << sz << " errno=" << errno;

			if (errno == EAGAIN) {
				// Just nothing to see here
				ret = 0;
			} else {
				ret = -1;
			}
			break;
		} else {
			ret = 1;
		}

		DEBUG("received ") << sz << " bytes";

		// Process data
		for (uint32_t i=0; i<sz; i++) {
//			fprintf(stdout, "[%d] Process char %c\n", m_msg_state, tmp[i]);
			switch (m_msg_state) {
			case 0: { // Waiting for a header
				if (tmp[i] == HEADER_PREFIX.at(m_msgbuf_idx)) {
					m_msgbuf_idx++;
				} else {
					m_msgbuf_idx = 0;
				}
				if (m_msgbuf_idx == HEADER_PREFIX.size()) {
					m_msgbuf_idx = 0;
					m_msg_state = 1;
				}
			} break;

			case 1: { // Collecting length up to first '\n'
				if (m_msgbuf_idx == 0 && isspace(tmp[i])) {
					// Skip leading whitespace
				} else {
					DEBUG("State 1: append ") << tmp[i];
					msgbuf_append(tmp[i]);
					if (isspace(tmp[i])) {
						msgbuf_append(0);
						DEBUG("header=") << m_msgbuf;
						m_msg_length = strtoul(m_msgbuf, 0, 10);
						DEBUG("len=") << m_msg_length;
						// Reset the buffer to collect the payload
						m_msgbuf_idx = 0;
						m_msg_state = 2;
					}
				}

			} break;

			case 2: { // Collecting body data
				if (m_msgbuf_idx == 0 && isspace(tmp[i])) {
					// Skip leading whitespace
				} else {
					msgbuf_append(tmp[i]);
					if (m_msgbuf_idx >= m_msg_length) {
						msgbuf_append(0);
						DEBUG("Received message: \"") << m_msgbuf << "\"";
						nlohmann::json msg;
						try {
							msg = nlohmann::json::parse(m_msgbuf);

							// Is this a request or a response?
							if (msg.contains("method")) {
								// TODO: Request or notify
								IParamValMapUP params(m_param_f.mkT<IParamValMap>(msg["params"]));

								intptr_t id = -1;

								if (msg.contains("id")) {
									id = msg["id"];
								}

								DEBUG_ENTER("m_req_f");
								m_outstanding += 1;
								m_req_f(msg["method"], id, params.get());
								DEBUG_LEAVE("m_req_f");
							} else {
								// TODO: Response
								IParamValMapUP result;
								IParamValMapUP error;
								ParamValIntUP id = ParamValInt::mk(msg["id"]);
								DEBUG_ENTER("m_rsp_f");
								if (msg.contains("result")) {
									result = IParamValMapUP(m_param_f.mkT<IParamValMap>(msg["result"]));
								} else if (msg.contains("error")) {
									error = IParamValMapUP(m_param_f.mkT<IParamValMap>(msg["error"]));
								}
								m_rsp_f(id->val_s(), result.get(), error.get());
								m_outstanding -= 1;
								DEBUG_LEAVE("m_rsp_f");
							}
						} catch (const std::exception &e) {
							fprintf(stdout, "Failed to parse msg \"%s\" %s\n",
									m_msgbuf, e.what());
						}
						m_msg_state = 0;
						m_msgbuf_idx = 0;
					}
				}
			} break;

			default: {
				m_msgbuf_idx = 0;
				m_msg_state = 0;
			}
			}
		}
	} while (ret == 0 || timeout_ms > 0);

	DEBUG_LEAVE("poll return ret=") << ret << "," << m_outstanding;
	if (ret == -1) {
		return -1;
	} else {
		return (ret || m_outstanding > 0);
	}
}
#endif /* UNDEFINED */

int32_t TransportJsonSocket::await_msg() {
	int32_t ret = 0;
	char tmp[1024];

	int32_t have_msg = 0;
	do {
		int32_t timeout_ms = 1000;

		// Select to wait on data
		int retval;
		do {
			fd_set rfds;
			struct timeval tv;

			FD_ZERO(&rfds);
			FD_SET(m_socket, &rfds);

			tv.tv_sec = timeout_ms/1000;
			tv.tv_usec = (timeout_ms%1000)*1000;

			retval = select(m_socket+1, &rfds, 0, 0, &tv);

		} while (retval == 0 || (retval == -1 && errno == EAGAIN));

		if (retval == -1) {
			ret = -1;
			break;
		}

		DEBUG_ENTER("poll recv");
		int32_t sz = ::recv(m_socket, tmp, sizeof(tmp), 0);
		DEBUG_LEAVE("poll recv");

		if (sz == -1 && errno != EAGAIN) {
			ret = -1;
		} else if (sz > 0) {
			ret = process_data(tmp, sz);
		} else if (sz == 0) {
			ret = -1;
		}
	} while (ret == 0);

	return ret;
}

intptr_t TransportJsonSocket::send_req(
		const std::string		&method,
		intptr_t				id,
		IParamValMap			*params) {
	DEBUG_ENTER("send_req");
	char tmp[64];
	nlohmann::json msg;
	msg["json-rpc"] = "2.0";
	msg["method"] = method;
	msg["id"] = id;
	msg["params"] = m_param_f.mk(params);

	std::string body = msg.dump();

//	DEBUG("body=\"") << body << " len=" << body.size();
	sprintf(tmp, "Content-Length: %d\r\n\r\n", body.size());

	m_outstanding += 1;

	int ret;

	if ((ret=::send(m_socket, tmp, strlen(tmp), MSG_NOSIGNAL)) == -1) {
		fprintf(stdout, "error status\n");
		return ret;
	}

	if ((ret=::send(m_socket, body.c_str(), body.size(), MSG_NOSIGNAL)) == -1) {
		fprintf(stdout, "error status\n");
		return ret;
	}

	// Ownership transfers back to us
	delete params;

	DEBUG_LEAVE("send_req");

	return 0;
}

int32_t TransportJsonSocket::send_notify(
		const std::string		&method,
		IParamValMap			*params) {
	DEBUG_ENTER("send_notify\n");
	char tmp[64];
	nlohmann::json msg;
	msg["json-rpc"] = "2.0";
	msg["method"] = method;
	msg["params"] = m_param_f.mk(params);

	int32_t ret = 0;

	std::string body = msg.dump();
//	DEBUG("body=\"") << body << "\" len=" << body.size();
	sprintf(tmp, "Content-Length: %d\r\n\r\n", body.size());

	::send(m_socket, tmp, strlen(tmp), 0);
	::send(m_socket, body.c_str(), body.size(), 0);

	delete params;

	DEBUG_LEAVE("send_notify");
	return ret;
}

int32_t TransportJsonSocket::send_rsp(
		intptr_t				id,
		IParamValMap			*result,
		IParamValMap			*error) {
	DEBUG_ENTER("send_rsp");
	char tmp[64];
	nlohmann::json msg;
	msg["json-rpc"] = "2.0";
	msg["id"] = id;
	if (result) {
		msg["result"] = m_param_f.mk(result);
		delete result;
	} else {
		msg["error"]  = m_param_f.mk(error);
		delete error;
	}

	std::string body = msg.dump();

//	DEBUG("body=\"") << body << "\" len=" << body.size();
	sprintf(tmp, "Content-Length: %d\r\n\r\n", body.size());

	int32_t ret = 0;

	if ((ret=::send(m_socket, tmp, strlen(tmp), MSG_NOSIGNAL)) == -1) {
		return ret;
	}
	if ((ret=::send(m_socket, body.c_str(), body.size(), MSG_NOSIGNAL)) == -1) {
		return ret;
	}

	m_outstanding -= 1;

	DEBUG_LEAVE("send_rsp");
	return ret;
}

IParamValBool *TransportJsonSocket::mkValBool(bool val) {
	return ParamValBool::mk(val).release();
}

IParamValInt *TransportJsonSocket::mkValIntU(uint64_t val, int32_t width) {
	return ParamValInt::mk(val).release();
}

IParamValInt *TransportJsonSocket::mkValIntS(int64_t val, int32_t width) {
	return ParamValInt::mk(val).release();
}

IParamValMap *TransportJsonSocket::mkValMap() {
	return ParamValMap::mk().release();
}

IParamValStr *TransportJsonSocket::mkValStr(const std::string &val) {
	return ParamValStr::mk(val).release();
}

IParamValVec *TransportJsonSocket::mkValVec() {
	return ParamValVec::mk().release();
}

int32_t TransportJsonSocket::process_data(char *data, uint32_t sz) {
	int32_t ret = 0;

	// Process data
	for (uint32_t i=0; i<sz; i++) {
		switch (m_msg_state) {
		case 0: { // Waiting for a header
			if (data[i] == HEADER_PREFIX.at(m_msgbuf_idx)) {
				m_msgbuf_idx++;
			} else {
				m_msgbuf_idx = 0;
			}
			if (m_msgbuf_idx == HEADER_PREFIX.size()) {
				m_msgbuf_idx = 0;
				m_msg_state = 1;
			}
		} break;

		case 1: { // Collecting length up to first '\n'
			if (m_msgbuf_idx == 0 && isspace(data[i])) {
				// Skip leading whitespace
			} else {
//				DEBUG("State 1: append ") << data[i];
				msgbuf_append(data[i]);
				if (isspace(data[i])) {
					msgbuf_append(0);
//					DEBUG("header=") << m_msgbuf;
					m_msg_length = strtoul(m_msgbuf, 0, 10);
//					DEBUG("len=") << m_msg_length;
					// Reset the buffer to collect the payload
					m_msgbuf_idx = 0;
					m_msg_state = 2;
				}
			}

		} break;

		case 2: { // Collecting body data
			if (m_msgbuf_idx == 0 && isspace(data[i])) {
				// Skip leading whitespace
			} else {
				msgbuf_append(data[i]);
				if (m_msgbuf_idx >= m_msg_length) {
					msgbuf_append(0);
					DEBUG("Received message %s", m_msgbuf);
					nlohmann::json msg;
					try {
						msg = nlohmann::json::parse(m_msgbuf);

						// We received a message
						ret = 1;

						// Is this a request or a response?
						if (msg.contains("method")) {
							// TODO: Request or notify
							IParamValMapUP params(m_param_f.mkT<IParamValMap>(msg["params"]));

							intptr_t id = -1;

							if (msg.contains("id")) {
								id = msg["id"];
							}

							DEBUG_ENTER("m_req_f");
							m_outstanding += 1;
							// TODO:
							m_req_f(msg["method"], id, params.get());
							DEBUG_LEAVE("m_req_f");
						} else {
							// TODO: Response
							IParamValMapUP result;
							IParamValMapUP error;
							ParamValIntUP id = ParamValInt::mk(msg["id"]);
							DEBUG_ENTER("m_rsp_f");
							if (msg.contains("result")) {
								result = IParamValMapUP(m_param_f.mkT<IParamValMap>(msg["result"]));
							} else if (msg.contains("error")) {
								error = IParamValMapUP(m_param_f.mkT<IParamValMap>(msg["error"]));
							}

							m_rsp_f(id->val_s(), result.get(), error.get());
							m_outstanding -= 1;
							DEBUG_LEAVE("m_rsp_f");
						}
					} catch (const std::exception &e) {
						fprintf(stdout, "Failed to parse msg \"%s\" %s\n",
								m_msgbuf, e.what());
					}
					m_msg_state = 0;
					m_msgbuf_idx = 0;
				}
			}
		} break;

		default: {
			m_msgbuf_idx = 0;
			m_msg_state = 0;
		}
		}
	}

	return ret;
}

void TransportJsonSocket::msgbuf_resize_append(char c) {
	// Confirm that we need to resize
	if (m_msgbuf_idx+1 >= m_msgbuf_max) {
		// yup, must resize
		char *tmp = m_msgbuf;
		m_msgbuf = new char[m_msgbuf_max+1024];

		memcpy(m_msgbuf, tmp, m_msgbuf_idx);
		m_msgbuf_max += 1024;

		delete [] tmp;
	}

	m_msgbuf[m_msgbuf_idx++] = c;
}

const std::string	TransportJsonSocket::HEADER_PREFIX = "Content-Length: ";

} /* namespace tblink_rpc_core */

